dot/core/messages.go

```diff
@@ -30,16 +30,23 @@ func (s *Service) validateTransaction(peerID peer.ID, head *types.Header, rt run
 
 	// validate each transaction
 	externalExt := types.Extrinsic(append([]byte{byte(types.TxnExternal)}, tx...))
-	validity, err = rt.ValidateTransaction(externalExt)
+	validity, txnValidityErr, err := rt.ValidateTransaction(externalExt)
```

Comment:
You could bundle that txnValidityErr inside an `error`.

To do so, the txnValidityErr type must implement the `error` interface:

```go
type error interface {
	Error() string
}
```

so just add that method `Error() string` to `txnvalidity.TransactionValidityError`.

And then in this part of the code, use something like:

```go
var txnValidityErr *txnvalidity.TransactionValidityError
if errors.As(err, &txnValidityErr) {
    // handle txnValidityErr
}
```

Reply:
@qdm12 So would this be able to check the type of the VDT? In this case txnValidityError is not an error type, and is used as a work around to type assertion.

Reply:
Not necessarily, but you can have it `txnvalidity.TransactionValidityError` as an `error` nonetheless.

Also I'm wondering in this case, do we need VDT? Or are just trying to have two error types in Go code (without scale codec)?

Reply:
I have to be able to do a type check on the VDT in the code to handle peer reporting correctly. See here https://github.com/ChainSafe/gossamer/blob/86a2e76b7675d42b275036b2e7a39f48d8a737ef/dot/core/messages.go#L39. I do not believe that there is a way in the code do something like error.DoesExistInVDT(), so I think this is needed still

Yes, it has to be a VDT. A transactionValidityError is a result of enums, so we have to use a VDT to encapsulate it.

I'm still not sure how I can make these changes without affecting the behavior.

Reply:
The third return param error will return if there's like an unmarshal error, or something is wrong with the wasmer instance.  Right now, when we unwrap a result it actually comes back as a `scale.WrappedErr` but I believe @jimjbrettj is explicitly returning a pointer to the Error case of the result.

Reply:
@qdm12 Talked with @timwu20 and what I did is implement the error interface while keeping the same type. Tim had said to implement as error as well, but if I do this I look access to the VDT methods. So now it implements both so its cleaner and I have access to all the needed functions.

When custom result types are implemented, I will change the signature of ValidateTransaction to return <CustomValidityResult, error> to make things less confusing and allow the caller to handle the result however it needs.